### PR TITLE
First try at implementing newly suggested ctor for providing tighter control over charset

### PR DESCRIPTION
**Pull request by brianmed**
There are three new ctor implementations.

All are from issue #17036.

```c#
public StringContent(string content, MediaTypeHeaderValue mediaType)
public StringContent(string content, Encoding encoding, MediaTypeHeaderValue mediaType)
public MediaTypeHeaderValue(string mediaType, string charSet)
```

Also, there are tests for the new APIs.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=63231) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=63231) before we can accept your contribution.<br/><br/>:x: brianmed [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=63231)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=63231) it.</sub>

File: src/libraries/System.Net.Http/ref/System.Net.Http.cs [hunk 1]
```diff
@@ -437,8 +437,10 @@ public partial class StreamContent : System.Net.Http.HttpContent
     public partial class StringContent : System.Net.Http.ByteArrayContent
     {
         public StringContent(string content) : base (default(byte[])) { }
+        public StringContent(string content, System.Net.Http.Headers.MediaTypeHeaderValue mediaType) : base (default(byte[])) { }
```
Comment (AraHaan) [inline src/libraries/System.Net.Http/ref/System.Net.Http.cs:440]: I thought that for reference code that ``base()`` should never be used.

Comment (brianmed) [reply src/libraries/System.Net.Http/ref/System.Net.Http.cs:440]: @AraHaan I'm actually clueless on that point.  What should be used?

Comment (AraHaan) [reply src/libraries/System.Net.Http/ref/System.Net.Http.cs:440]: Well, I see that it already uses ``base()`` so that file should be fine to do it in.

Comment (danmoseley) [reply src/libraries/System.Net.Http/ref/System.Net.Http.cs:440]: There is a markdown doc in this repo explaining how to update the ref files automatically.

Comment (brianmed) [reply src/libraries/System.Net.Http/ref/System.Net.Http.cs:440]: @danmoseley Thank you for letting me know.  After I follow the instructions in updating-ref-source.md I'm getting a large number of changes.  Many of those changes I did not do.

Comment (danmoseley) [reply src/libraries/System.Net.Http/ref/System.Net.Http.cs:440]: Are they just visual (folks didn't use the tool) or functional? Anyway, you can just revert unrelated changes.

Comment (brianmed) [reply src/libraries/System.Net.Http/ref/System.Net.Http.cs:440]: One is an issue I should fix.  StringContent's mediaType should be string and is currently string?.

There are also several visual differences and several differences that feel like they are functional.  For example, here are some Equals differences.   Should I attach a file of all the differences (from git diff)?

-        public bool Equals([System.Diagnostics.CodeAnalysis.NotNullWhen(true)] System.Net.Http.HttpMethod? other) { throw null; }
-        public override bool Equals([System.Diagnostics.CodeAnalysis.NotNullWhen(true)] object? obj) { throw null; }
+        public bool Equals([System.Diagnostics.CodeAnalysis.NotNullWhenAttribute(true)] System.Net.Http.HttpMethod? other) { throw null; }
+        public override bool Equals([System.Diagnostics.CodeAnalysis.NotNullWhenAttribute(true)] object? obj) { throw null; }

Comment (AraHaan) [reply src/libraries/System.Net.Http/ref/System.Net.Http.cs:440]: Perhaps the tool should be updated.

Comment (danmoseley) [reply src/libraries/System.Net.Http/ref/System.Net.Http.cs:440]: The ideal, I think, is that the ref files are completely and only generated output. cc @safern for whether that's correct.

Meantime I suggest to just include relevant changes.

Comment (brianmed) [reply src/libraries/System.Net.Http/ref/System.Net.Http.cs:440]: I just did a commit that hopefully updated the reference assembly correctly.

Comment (brianmed) [reply src/libraries/System.Net.Http/ref/System.Net.Http.cs:440]: @danmoseley Was my latest commit sufficient?

Comment (danmoseley) [reply src/libraries/System.Net.Http/ref/System.Net.Http.cs:440]: @brianmed looks right as far as I can tell, thanks. I guess @scalablecory plans to review now.

File: src/libraries/System.Net.Http/ref/System.Net.Http.cs [hunk 2]
```diff
@@ -437,8 +437,10 @@ public partial class StreamContent : System.Net.Http.HttpContent
...
         public StringContent(string content, System.Text.Encoding? encoding) : base (default(byte[])) { }
-        public StringContent(string content, System.Text.Encoding? encoding, string? mediaType) : base (default(byte[])) { }
+        public StringContent(string content, System.Text.Encoding? encoding, System.Net.Http.Headers.MediaTypeHeaderValue mediaTypeHeaderValue) : base (default(byte[])) { }
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/ref/System.Net.Http.cs:442]:
```suggestion
        public StringContent(string content, System.Text.Encoding? encoding, System.Net.Http.Headers.MediaTypeHeaderValue mediaType) : base (default(byte[])) { }
```

File: src/libraries/System.Net.Http/src/System/Net/Http/StringContent.cs [hunk 3]
```diff
@@ -14,23 +14,29 @@ public class StringContent : ByteArrayContent
...
+        public StringContent(string content, Encoding? encoding, string mediaType)
+            : this(content, encoding, new MediaTypeHeaderValue(mediaType, (encoding == null) ? DefaultStringEncoding.WebName : encoding.WebName))
         {
-            // Initialize the 'Content-Type' header with information provided by parameters.
-            MediaTypeHeaderValue headerValue = new MediaTypeHeaderValue((mediaType == null) ? DefaultMediaType : mediaType);
-            headerValue.CharSet = (encoding == null) ? HttpContent.DefaultStringEncoding.WebName : encoding.WebName;
+        }
 
-            Headers.ContentType = headerValue;
+        public StringContent(string content, Encoding? encoding, MediaTypeHeaderValue mediaTypeHeaderValue)
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/StringContent.cs:36]:
```suggestion
        public StringContent(string content, Encoding? encoding, MediaTypeHeaderValue mediaType)
```

File: src/libraries/System.Net.Http/src/System/Net/Http/StringContent.cs [hunk 4]
```diff
@@ -14,23 +14,29 @@ public class StringContent : ByteArrayContent
...
         public StringContent(string content, Encoding? encoding)
-            : this(content, encoding, null)
+            : this(content, encoding, DefaultMediaType)
         {
         }
 
-        public StringContent(string content, Encoding? encoding, string? mediaType)
-            : base(GetContentByteArray(content, encoding))
+        public StringContent(string content, Encoding? encoding, string mediaType)
+            : this(content, encoding, new MediaTypeHeaderValue(mediaType, (encoding == null) ? DefaultStringEncoding.WebName : encoding.WebName))
```
Comment (scalablecory) [inline src/libraries/System.Net.Http/src/System/Net/Http/StringContent.cs:32]:
```suggestion
            : this(content, encoding, new MediaTypeHeaderValue(mediaType, (encoding ?? DefaultStringEncoding).WebName))
```

**Comment by karelz** [comment]
Hello @brianmed, are you back from vacation? Do you have rough ETA when you think you can get to the remaining feedback? Thanks!

File: src/libraries/System.Net.Http/ref/System.Net.Http.cs (hunk 1 above)
Comment (brianmed) [reply src/libraries/System.Net.Http/ref/System.Net.Http.cs:440]: @danmoseley Just updated.  Sorry it took a few days.

**Review by scalablecory — APPROVED** [review]
lgtm.

**Review by stephentoub — APPROVED** [review]
thanks!

**Pull request merged**